File: xflaim/src/java/wrapper/xflaim/DataVector.java
//------------------------------------------------------------------------------
// Desc:	Data Vector
// Tabs:	3
//
// Copyright (c) 2003-2007 Novell, Inc. All Rights Reserved.
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; version 2.1
// of the License.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Library Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, contact Novell, Inc.
//
// To contact Novell about this file by physical or electronic mail, 
// you may find current contact information at www.novell.com.
//
// $Id$
//------------------------------------------------------------------------------

package xflaim;

/**
 * This class implements an interface to the XFlaim IF_DataVector class.
 */
public class DataVector
{
	long		m_this;

	/**
	 * Constructor for the DataVector object.  This object provides access to
	 * the XFlaim IF_DataVector interface.  All methods defined by the
	 * IF_DataVector interace are accessible through this Java object.
	 * 
	 * @param lRef A reference to a C++ IF_DataVector object
	 */
	public DataVector(
		long		lRef)
	{
		super();
		m_this = lRef;
	}
	
	public long getThis()
	{
		return m_this;
	}
	
	/**
	 * Finalizer method, used to ensure that we release the actual C++ object.
	 */
	public void finalize()
	{
		if (m_this != 0)
		{
			_release( m_this);
			m_this = 0;
		}
	}
	
	/**
	 * Set the document ID of the data vector.
	 * @param lDocId Document ID for the data vector.
	 * @throws XFlaimException
	 */
	public void setDocumentID(
		long		lDocId) throws XFlaimException
	{
		_setDocumentId( m_this, lDocId);
	}
	
	/**
	 * Set the node ID of an element in the data vector. 
	 * @param iElementNumber Element whose node ID is to be set.
	 * @param lID Node ID value that is to be set.
	 * @throws XFlaimException
	 */
	public void setID(
		int			iElementNumber,
		long			lID) throws XFlaimException
	{
		_setID( m_this, iElementNumber, lID);
	}
	
	/**
	 * Set the name ID of an element in the data vector.
	 * @param iElementNumber Element whose name ID is to be set.
	 * @param iNameId Name ID that is to be set into the element.
	 * @param bIsAttr A boolean flag that indicates whether or not this
	 * name ID is an attribute or an element.  A value of true means it is
	 * an attribute, false means it is an element.
	 * @param bIsData A boolean flag that indicates whether or not the
	 * element is a data component as opposed to a key component.
	 * @throws XFlaimException
	 */
	public void setNameId(
		int			iElementNumber,
		int			iNameId,
		boolean		bIsAttr,
		boolean		bIsData) throws XFlaimException
	{
		_setNameId( m_this, iElementNumber, iNameId, bIsAttr, bIsData);
	}

	/**
	 * Set the value of an element in the data vector to a long value.
	 * @param iElementNumber Element whose value is to be set.
	 * @param lNum The long value that is to be set into the element.
	 * @throws XFlaimException
	 */
	public void setLong(
		int			iElementNumber,
		long			lNum) throws XFlaimException
	{
		_setLong( m_this, iElementNumber, lNum);
	}

	/**
	 * Set the value of an element in the data vector to a String value.
	 * @param iElementNumber Element whose value is to be set.
	 * @param sValue The string value that is to be set into the element.
	 * @throws XFlaimException
	 */
	public void setString(
		int			iElementNumber,
		String		sValue) throws XFlaimException
	{
		_setString( m_this, iElementNumber, sValue);
	}
	
	/**
	 * Set the value of an element in the data vector to a binary value.
	 * @param iElementNumber Element whose value is to be set.
	 * @param Value The binary value that is to be set into the element.
	 * @throws XFlaimException
	 */
	public void setBinary(
		int			iElementNumber,
		byte[]		Value) throws XFlaimException
	{
		_setBinary( m_this, iElementNumber, Value);
	}
	
	/**
	 * Set a flag of an element in the data vector to indicate that the value
	 * is right truncated.
	 * @param iElementNumber Element whose "right truncated" flag is to be set.
	 * @throws XFlaimException
	 */
	public void setRightTruncated(
		int		iElementNumber) throws XFlaimException
	{
		_setRightTruncated( m_this, iElementNumber);
	}

	/**
	 * Set a flag of an element in the data vector to indicate that the value
	 * is left truncated.
	 * @param iElementNumber Element whose "left truncated" flag is to be set.
	 * @throws XFlaimException
	 */
	public void setLeftTruncated(
		int		iElementNumber) throws XFlaimException
	{
		_setLeftTruncated( m_this, iElementNumber);
	}

	/**
	 * Clear the "right truncated" flag of an element in the data vector.
	 * @param iElementNumber Element whose "right truncated" flag is to be cleared.
	 * @throws XFlaimException
	 */
	public void clearRightTruncated(
		int		iElementNumber) throws XFlaimException
	{
		_clearRightTruncated( m_this, iElementNumber);
	}

	/**
	 * Clear the "left truncated" flag of an element in the data vector.
	 * @param iElementNumber Element whose "left truncated" flag is to be cleared.
	 * @throws XFlaimException
	 */
	public void clearLeftTruncated(
		int		iElementNumber) throws XFlaimException
	{
		_clearLeftTruncated( m_this, iElementNumber);
	}

	/**
	 * Determine if an element in the data vector is "right truncated."
	 * @param iElementNumber Element whose "right truncated" flag is to be checked.
	 * @return Returns flag indicating whether or not the specified element is
	 * "right truncated."
	 * @throws XFlaimException
	 */
	public void isRightTruncated(
		int		iElementNumber) throws XFlaimException
	{
		_isRightTruncated( m_this, iElementNumber);
	}

	/**
	 * Determine if an element in the data vector is "left truncated."
	 * @param iElementNumber Element whose "left truncated" flag is to be checked.
	 * @return Returns flag indicating whether or not the specified element is
	 * "left truncated."
	 * @throws XFlaimException
	 */
	public void isLeftTruncated(
		int		iElementNumber) throws XFlaimException
	{
		_isLeftTruncated( m_this, iElementNumber);
	}

	/**
	 * Get the Document ID for the data vector.
	 * @return Returns document ID.
	 * @throws XFlaimException
	 */
	public long getDocumentID() throws XFlaimException
	{
		return _getDocumentID( m_this);
	}

	/**
	 * Get the node ID for an element in the data vector.
	 * @param iElementNumber Element whose node ID is to be returned.
	 * @return Returns element's node ID.
	 * @throws XFlaimException
	 */
	public long getID(
		int		iElementNumber) throws XFlaimException
	{
		return _getID( m_this, iElementNumber);
	}

	/**
	 * Get the name ID for an element in the data vector.
	 * @param iElementNumber Element whose name ID is to be returned.
	 * @return Returns element's name ID.
	 * @throws XFlaimException
	 */
	public int getNameId(
		int		iElementNumber) throws XFlaimException
	{
		return _getNameId( m_this, iElementNumber);
	}

	/**
	 * Determine the name ID for an element in the data vector is an attribute
	 * name ID or an element name ID
	 * @param iElementNumber Element whose name ID is to be tested.
	 * @return Returns flag indicating if the name ID is an element name
	 * (returns false) or an attribute name (returns true).
	 * @throws XFlaimException
	 */
	public boolean isAttr(
		int		iElementNumber) throws XFlaimException
	{
		return _isAttr( m_this, iElementNumber);
	}

	/**
	 * Determine an element in the data vector is a data component.
	 * @param iElementNumber Element to be tested.
	 * @return Returns flag indicating if the element is a data component.
	 * @throws XFlaimException
	 */
	public boolean isDataComponent(
		int		iElementNumber) throws XFlaimException
	{
		return _isDataComponent( m_this, iElementNumber);
	}

	/**
	 * Determine an element in the data vector is a key component.
	 * @param iElementNumber Element to be tested.
	 * @return Returns flag indicating if the element is a key component.
	 * @throws XFlaimException
	 */
	public boolean isKeyComponent(
		int		iElementNumber) throws XFlaimException
	{
		return _isKeyComponent( m_this, iElementNumber);
	}

	/**
	 * Get the length of the data value of an element in the data vector.
	 * @param iElementNumber Element whose data value length is to be returned.
	 * @return Returns element's data value length.
	 * @throws XFlaimException
	 */
	public int getDataLength(
		int		iElementNumber) throws XFlaimException
	{
		return _getDataLength( m_this, iElementNumber);
	}

	/**
	 * Get the type of the data value of an element in the data vector.
	 * @param iElementNumber Element whose data value type is to be returned.
	 * @return Returns element's data value type.  This will be one of the
	 * values of {@link xflaim.FlmDataType FlmDataType}.
	 * @throws XFlaimException
	 */
	public int getDataType(
		int		iElementNumber) throws XFlaimException
	{
		return _getDataType( m_this, iElementNumber);
	}

	/**
	 * Get the value of an element in the data vector as a long value.
	 * @param iElementNumber Element whose data value is to be returned.
	 * @return Returns element's data value as a long.
	 * @throws XFlaimException
	 */
	public long getLong(
		int		iElementNumber) throws XFlaimException
	{
		return _getLong( m_this, iElementNumber);
	}
	
	/**
	 * Get the value of an element in the data vector as a String.
	 * @param iElementNumber Element whose data value is to be returned.
	 * @return Returns element's data value as a string.
	 * @throws XFlaimException
	 */
	public String getString(
		int		iElementNumber) throws XFlaimException
	{
		return _getString( m_this, iElementNumber);
	}

	/**
	 * Get the value of an element in the data vector as a binary value.
	 * @param iElementNumber Element whose data value is to be returned.
	 * @return Returns element's data value as a binary value.
	 * @throws XFlaimException
	 */
	public byte[] getBinary(
		int		iElementNumber) throws XFlaimException
	{
		return _getBinary( m_this, iElementNumber);
	}

	/**
	 * Return a buffer that is an index key for the elements in the vector.  The
	 * key is generated using the definition of the index from the specified
	 * database.
	 * @param jDb Database containing the index definition for which we want to
	 * generate a key.
	 * @param iIndexNum Index number for which we want to generate a key.
	 * @param bOutputIds Flag that specifies whether or not node ids and
	 * document ids in the data vector are to be included in the generated key.
	 * @return Returns the generated index key.
	 * @throws XFlaimException
	 */
	public byte[] outputKey(
		Db				jDb,
		int			iIndexNum,
		boolean		bOutputIds) throws XFlaimException
	{
		return _outputKey( m_this, jDb.m_this, iIndexNum, bOutputIds);
	}

	/**
	 * Return a buffer that contains the data component part of an index key.
	 * the buffer is composed from the elements in the data vector that are
	 * data components.
	 * @param jDb Database containing the index definition for which we want to
	 * generate the data component.
	 * @param iIndexNum Index number for which we want to generate the data
	 * component.
	 * @return Returns the generated data component part of an index key.
	 * @throws XFlaimException
	 */
	public byte[] outputData(
		Db			jDb,
		int		iIndexNum) throws XFlaimException
	{
		return _outputData( m_this, jDb.m_this, iIndexNum);
	}

	/**
	 * Populate the data vector's key components using a buffer that contains
	 * an index key.
	 * @param jDb Database containing the index definition that is to be used to
	 * parse through the key buffer to determine the key components.
	 * @param iIndexNum Index number to be used to parse through the key buffer
	 * to determine the key components.
	 * @param Key Buffer containing the index key to be parsed.
	 * @throws XFlaimException
	 */
	public void inputKey(
		Db				jDb,
		int			iIndexNum,
		byte[]		Key) throws XFlaimException
	{
		_inputKey( m_this, jDb.m_this, iIndexNum, Key);
	}

	/**
	 * Populate the data vector's data components using a buffer that contains
	 * an index key's data components.
	 * @param jDb Database containing the index definition that is to be used to
	 * parse through the data buffer to determine the data components.
	 * @param iIndexNum Index number to be used to parse through the data buffer
	 * to determine the data components.
	 * @param Data Buffer containing the index key's data components to be parsed.
	 * @throws XFlaimException
	 */
	public void inputData(
		Db				jDb,
		int			iIndexNum,
		byte[]		Data) throws XFlaimException
	{
		_inputData( m_this, jDb.m_this, iIndexNum, Data);
	}
	
	/**
	 * Reset the contents of the data vector.
	 * @throws XFlaimException
	 */
	public void reset() throws XFlaimException
	{
		_reset( m_this);
	}
	
// PRIVATE METHODS

	private native void _release( 
		long 		lThis);
	
	private native void _setDocumentId(
		long		lThis,
		long		lDocId) throws XFlaimException;

	private native void _setID(
		long		lThis,
		int		iElementNumber,
		long		lID) throws XFlaimException;
		
	private native void _setNameId(
		long		lThis,
		int		iElementNumber,
		int		iNameId,
		boolean	bIsAttr,
		boolean	bIsData) throws XFlaimException;
		
	private native void _setLong(
		long		lThis,
		int		iElementNumber,
		long		lNum) throws XFlaimException;
		
	private native void _setString(
		long		lThis,
		int		iElementNumber,
		String	sValue) throws XFlaimException;
	
	private native void _setBinary(
		long		lThis,
		int		iElementNumber,
		byte[]	Value) throws XFlaimException;
	
	private native void _setRightTruncated(
		long		lThis,
		int		iElementNumber) throws XFlaimException;

	private native void _setLeftTruncated(
		long		lThis,
		int		iElementNumber) throws XFlaimException;

	private native void _clearRightTruncated(
		long		lThis,
		int		iElementNumber) throws XFlaimException;

	private native void _clearLeftTruncated(
		long		lThis,
		int		iElementNumber) throws XFlaimException;
	
	private native boolean _isRightTruncated(
		long		lThis,
		int		iElementNumber) throws XFlaimException;
	
	private native boolean _isLeftTruncated(
		long		lThis,
		int		iElementNumber) throws XFlaimException;
	
	private native long _getDocumentID(
		long		lThis) throws XFlaimException;
	
	private native long _getID(
		long		lThis,
		int		iElementNumber) throws XFlaimException;
		
	private native int _getNameId(
		long		lThis,
		int		iElementNumber) throws XFlaimException;

	private native boolean _isAttr(
		long		lThis,
		int		iElementNumber) throws XFlaimException;

	private native boolean _isDataComponent(
		long		lThis,
		int		iElementNumber) throws XFlaimException;

	private native boolean _isKeyComponent(
		long		lThis,
		int		iElementNumber) throws XFlaimException;

	private native int _getDataLength(
		long		lThis,
		int		iElementNumber) throws XFlaimException;

	private native int _getDataType(
		long		lThis,
		int		iElementNumber) throws XFlaimException;

	private native long _getLong(
		long		lThis,
		int		iElementNumber) throws XFlaimException;

	private native String _getString(
		long		lThis,
		int		iElementNumber) throws XFlaimException;

	private native byte[] _getBinary(
		long		lThis,
		int		iElementNumber) throws XFlaimException;

	private native byte[] _outputKey(
		long		lThis,
		long		ljDbRef,
		int		iIndexNum,
		boolean	bOutputIds) throws XFlaimException;

	private native byte[] _outputData(
		long		lThis,
		long		ljDbRef,
		int		iIndexNum) throws XFlaimException;

	private native void _inputKey(
		long		lThis,
		long		ljDbRef,
		int		iIndexNum,
		byte[]	Key) throws XFlaimException;

	private native void _inputData(
		long		lThis,
		long		ljDbRef,
		int		iIndexNum,
		byte[]	Data) throws XFlaimException;

	private native void _reset(
		long		lThis) throws XFlaimException;
}

File: xflaim/src/cs/DataVector.cpp
//------------------------------------------------------------------------------
// Desc: Native C routines to support C# DataVector class
// Tabs:	3
//
//	Copyright (c) 2006-2007 Novell, Inc. All Rights Reserved.
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; version 2.1
// of the License.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Library Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, contact Novell, Inc.
//
// To contact Novell about this file by physical or electronic mail, 
// you may find current contact information at www.novell.com.
//
// $Id$
//------------------------------------------------------------------------------

#include "xflaim.h"

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC void XFLAPI xflaim_DataVector_Release(
	IF_DataVector *	pDataVector)
{
	if (pDataVector)
	{
		pDataVector->Release();
	}
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC void XFLAPI xflaim_DataVector_setDocumentID(
	IF_DataVector *	pDataVector,
	FLMUINT64			ui64DocumentID)
{
	pDataVector->setDocumentID( ui64DocumentID);
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_DataVector_setID(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber,
	FLMUINT64			ui64ID)
{
	return( pDataVector->setID( (FLMUINT)ui32ElementNumber, ui64ID));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_DataVector_setNameId(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber,
	FLMUINT32			ui32NameId,
	FLMBOOL				bIsAttr,
	FLMBOOL				bIsData)
{
	return( pDataVector->setNameId( (FLMUINT)ui32ElementNumber,
					(FLMUINT)ui32NameId, bIsAttr, bIsData));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_DataVector_setULong(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber,
	FLMUINT64			ui64Value)
{
	return( pDataVector->setUINT64( (FLMUINT)ui32ElementNumber, ui64Value));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_DataVector_setLong(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber,
	FLMINT64				i64Value)
{
	return( pDataVector->setINT64( (FLMUINT)ui32ElementNumber, i64Value));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_DataVector_setUInt(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber,
	FLMUINT32			ui32Value)
{
	return( pDataVector->setUINT( (FLMUINT)ui32ElementNumber, (FLMUINT)ui32Value));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_DataVector_setInt(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber,
	FLMINT32				i32Value)
{
	return( pDataVector->setINT( (FLMUINT)ui32ElementNumber, (FLMINT)i32Value));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_DataVector_setString(
	IF_DataVector *		pDataVector,
	FLMUINT32				ui32ElementNumber,
	const FLMUNICODE *	puzValue)
{
	return( pDataVector->setUnicode( (FLMUINT)ui32ElementNumber, puzValue));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_DataVector_setBinary(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber,
	const void *		pvValue,
	FLMUINT32			ui32Len)
{
	return( pDataVector->setBinary( (FLMUINT)ui32ElementNumber, pvValue, (FLMUINT)ui32Len));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC void XFLAPI xflaim_DataVector_setRightTruncated(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber)
{
	pDataVector->setRightTruncated( (FLMUINT)ui32ElementNumber);
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC void XFLAPI xflaim_DataVector_setLeftTruncated(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber)
{
	pDataVector->setLeftTruncated( (FLMUINT)ui32ElementNumber);
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC void XFLAPI xflaim_DataVector_clearRightTruncated(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber)
{
	pDataVector->clearRightTruncated( (FLMUINT)ui32ElementNumber);
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC void XFLAPI xflaim_DataVector_clearLeftTruncated(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber)
{
	pDataVector->clearLeftTruncated( (FLMUINT)ui32ElementNumber);
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC FLMBOOL XFLAPI xflaim_DataVector_isRightTruncated(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber)
{
	return( pDataVector->isRightTruncated( (FLMUINT)ui32ElementNumber));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC FLMBOOL XFLAPI xflaim_DataVector_isLeftTruncated(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber)
{
	return( pDataVector->isLeftTruncated( (FLMUINT)ui32ElementNumber));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC FLMUINT64 XFLAPI xflaim_DataVector_getDocumentID(
	IF_DataVector *	pDataVector)
{
	return( pDataVector->getDocumentID());
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC FLMUINT64 XFLAPI xflaim_DataVector_getID(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber)
{
	return( pDataVector->getID( (FLMUINT)ui32ElementNumber));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC FLMUINT32 XFLAPI xflaim_DataVector_getNameId(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber)
{
	return( (FLMUINT32)pDataVector->getNameId( (FLMUINT)ui32ElementNumber));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC FLMBOOL XFLAPI xflaim_DataVector_isAttr(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber)
{
	return( pDataVector->isAttr( (FLMUINT)ui32ElementNumber));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC FLMBOOL XFLAPI xflaim_DataVector_isDataComponent(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber)
{
	return( pDataVector->isDataComponent( (FLMUINT)ui32ElementNumber));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC FLMBOOL XFLAPI xflaim_DataVector_isKeyComponent(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber)
{
	return( pDataVector->isKeyComponent( (FLMUINT)ui32ElementNumber));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC FLMUINT32 XFLAPI xflaim_DataVector_getDataLength(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber)
{
	return( (FLMUINT32)pDataVector->getDataLength( (FLMUINT)ui32ElementNumber));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC FLMUINT32 XFLAPI xflaim_DataVector_getDataType(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber)
{
	return( (FLMUINT32)pDataVector->getDataType( (FLMUINT)ui32ElementNumber));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_DataVector_getULong(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber,
	FLMUINT64 *			pui64Value)
{
	return( pDataVector->getUINT64( (FLMUINT)ui32ElementNumber, pui64Value));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_DataVector_getLong(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber,
	FLMINT64 *			pi64Value)
{
	return( pDataVector->getINT64( (FLMUINT)ui32ElementNumber, pi64Value));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_DataVector_getUInt(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber,
	FLMUINT32 *			pui32Value)
{
	RCODE		rc;
	FLMUINT	uiValue;
	
	rc = pDataVector->getUINT( (FLMUINT)ui32ElementNumber, &uiValue);
	*pui32Value = (FLMUINT32)uiValue;
	return( rc);
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_DataVector_getInt(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber,
	FLMINT32 *			pi32Value)
{
	RCODE		rc;
	FLMINT	iValue;
	
	rc = pDataVector->getINT( (FLMUINT)ui32ElementNumber, &iValue);
	*pi32Value = (FLMINT32)iValue;
	return( rc);
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_DataVector_getString(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber,
	FLMUNICODE **		ppuzValue)
{
	return( pDataVector->getUnicode( (FLMUINT)ui32ElementNumber, ppuzValue));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_DataVector_getBinary(
	IF_DataVector *	pDataVector,
	FLMUINT32			ui32ElementNumber,
	FLMUINT32			ui32Len,
	void *				pvValue)
{
	RCODE		rc = NE_XFLM_OK;
	FLMUINT	uiLength = (FLMUINT)ui32Len;
	
	if (RC_BAD( rc = pDataVector->getBinary( (FLMUINT)ui32ElementNumber,
		pvValue, &uiLength)))
	{
		goto Exit;
	}
	flmAssert( uiLength == (FLMUINT)ui32Len);

Exit:

	return( rc);
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_DataVector_outputKey(
	IF_DataVector *	pDataVector,
	IF_Db *				pDb,
	FLMUINT32			ui32IndexNum,
	FLMBOOL				bOutputIds,
	FLMBYTE *			pucKey,
	FLMINT32 *			pi32Len)
{
	RCODE		rc = NE_XFLM_OK;
	FLMUINT	uiLength;

	if (RC_BAD( rc = pDataVector->outputKey( pDb, (FLMUINT)ui32IndexNum,
			bOutputIds, pucKey, XFLM_MAX_KEY_SIZE, &uiLength)))
	{
		goto Exit;
	}
	*pi32Len = (FLMINT32)uiLength;
	
	
Exit:

	return( rc);
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_DataVector_outputData(
	IF_DataVector *	pDataVector,
	IF_Db *				pDb,
	FLMUINT32			ui32IndexNum,
	FLMBYTE *			pucData,
	FLMINT32				i32BufSize,
	FLMINT32 *			pi32Len)
{
	RCODE		rc = NE_XFLM_OK;
	FLMUINT	uiLength;

	if (RC_BAD( rc = pDataVector->outputData( pDb, (FLMUINT)ui32IndexNum,
			pucData, (FLMUINT)i32BufSize, &uiLength)))
	{
		goto Exit;
	}
	*pi32Len = (FLMINT32)uiLength;
	
Exit:

	return( rc);
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_DataVector_inputKey(
	IF_DataVector *	pDataVector,
	IF_Db *				pDb,
	FLMUINT32			ui32IndexNum,
	FLMBYTE *			pucKey,
	FLMINT32				i32KeyLen)
{
	return( pDataVector->inputKey( pDb, (FLMUINT)ui32IndexNum,
			pucKey, (FLMUINT)i32KeyLen));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC RCODE XFLAPI xflaim_DataVector_inputData(
	IF_DataVector *	pDataVector,
	IF_Db *				pDb,
	FLMUINT32			ui32IndexNum,
	FLMBYTE *			pucData,
	FLMINT32				i32DataLen)
{
	return( pDataVector->inputData( pDb, (FLMUINT)ui32IndexNum,
			pucData, (FLMUINT)i32DataLen));
}

/****************************************************************************
Desc:
****************************************************************************/
XFLXPC void XFLAPI xflaim_DataVector_reset(
	IF_DataVector *	pDataVector)
{
	pDataVector->reset();
}
